### PR TITLE
fix #25010: added midi generation for guitar bends

### LINKED ISSUE #25010: Bends are not exported in MIDI
### Issue type

Import/export issue

### Description with steps to reproduce

There are no bends in exported MIDI from the MU Studio, also grace notes and prebends are exported as separate regular notes (will paste the link for another issue)
Example score for export: 
[Bends_export example.mscz.zip](https://github.com/user-attachments/files/17227056/Bends_export.example.mscz.zip)

Steps to reproduce:
1. Open the attached score with bands (or any score with bends in it)
2. Export MIDI from the score
3. Open MIDI and check bends playback

### Supporting files, videos and screenshots

Actual result:
No bends in exported MIDI from the score
[Bends_export example.mid.zip](https://github.com/user-attachments/files/17227160/Bends_export.example.mid.zip)

Expected result:
Bends are exported

### What is the latest version of MuseScore Studio where this issue is present?

4.4.2

### Regression

No.

### Operating system

macOS 14

### Additional context

_No response_

### Checklist

- [X] This report follows the [guidelines](https://github.com/musescore/MuseScore/wiki/Reporting-bugs-and-issues) for reporting bugs and issues
- [X] I have verified that this issue has not been logged before, by searching the [issue tracker](https://github.com/musescore/MuseScore/issues) for similar issues
- [X] I have attached all requested files and information to this report
- [X] I have attempted to identify the root problem as concisely as possible, and have used minimal reproducible examples where possible

### PR DESCRIPTION
**Pull request by alexpavlov96**
Resolves: https://github.com/musescore/MuseScore/issues/25010

changes that affect other things in midi export than only `GuitarBend`:

1. Duration of note with slide after grace note:
`el->push_back(NoteEvent(0, onTime, slideOn - onTime, velocity, !note->tieBack()));`
example:
[grace+slide.gp.zip](https://github.com/user-attachments/files/17292804/grace%2Bslide.gp.zip)

2. Resetting pitchwheel to 0 as close to next note as possible.
```
if (tick - specs.mStep > pwEvent->first) {
    tickForPwReset = tick - specs.mStep;
} else {
    tickForPwReset = (tick + pwEvent->first) / 2;
}

```
With new implementation of bends we have 2 notes (instead of 1), so we skip sound ON of second note (adding more pitchbends to keep sound on the needed pitch), so if pitchbend should be dropped as close as possible to next note - not to have sound artifacts

Screenshot 2024-10-08 at 15 05 49

example:
[bend+note.mscz.zip](https://github.com/user-attachments/files/17292985/bend%2Bnote.mscz.zip)

### CODE REVIEW

File: src/engraving/compat/midi/compatmidirender.cpp [hunk 1]
```diff
@@ -932,7 +945,7 @@ void CompatMidiRender::createSlideOutNotePlayEvents(Note* note, NoteEventList* e
     int slideOn = NoteEvent::NOTE_LENGTH - totalSlideDuration;
     double velocity = !note->ghost() ? NoteEvent::DEFAULT_VELOCITY_MULTIPLIER : NoteEvent::GHOST_VELOCITY_MULTIPLIER;
     if (!hasTremolo) {
-        el->push_back(NoteEvent(0, onTime, slideOn, velocity, !note->tieBack()));
+        el->push_back(NoteEvent(0, onTime, slideOn - onTime, velocity, !note->tieBack()));
```
Comment (mikekirin) [inline src/engraving/compat/midi/compatmidirender.cpp:948]: I'm not 100% sure, but it seems NoteEvent m_len will always be 0 (slideOn-onTime). So this event makes no sense.

**Review by abariska — COMMENTED** [review]
1. Grace-bend should take its duration from previous not (not from the target one).
In this case it should be part of A note. In properties of the grace-bend diagram should reproduce actual previous not (A), which allows to make the grace-bend note as long as we want.
abariska 2024-10-11 at 17 08 22
[gracebend+bend+release.mscz.zip](https://github.com/user-attachments/files/17344425/gracebend%2Bbend%2Brelease.mscz.zip)

2. Bend after prebend ends on one point earlier that it should. 
Last pitch change should be at the begin of the beat.
abariska 2024-10-10 at 17 53 08
[prebend+bend.mscz.zip](https://github.com/user-attachments/files/17344419/prebend%2Bbend.mscz.zip)

3. Pitch of bends changes to less of pitch points (e.g. 8-9 for full bend instead of 10)

https://github.com/user-attachments/assets/40ec7d88-a392-4415-8d38-ec23fba6de55

4. 3 pitch points change instead 2 to be used for slight-bend.

File: src/engraving/compat/midi/compatmidirender.cpp [hunk 2]
```diff
@@ -495,7 +498,20 @@ void CompatMidiRender::updateGateTime(const Instrument* instrument, int& gateTim
...
 void CompatMidiRender::createGraceNotesPlayEvents(const Score* score, const Fraction& tick, Chord* chord, int& ontime, int& trailtime)
 {
-    std::vector<Chord*> gnb = chord->graceNotesBefore();
+    auto hasNoPrebendGrace = [] (Chord* chord) -> bool {
+        return std::any_of(chord->notes().begin(), chord->notes().end(), [](Note* n) {
+            return !n->isPreBendStart();
+        });
+    };
+
+    std::vector<Chord*> gnbNotExcludingPrebends = chord->graceNotesBefore();
```
Comment (RomanPudashkin) [inline src/engraving/compat/midi/compatmidirender.cpp:507]: You could call it with false to filter out the pre-bend note

File: src/engraving/compat/midi/compatmidirender.cpp [hunk 3]
```diff
@@ -1145,7 +1179,7 @@ int CompatMidiRender::totalTiedNoteTicks(Note* note)
     return total.ticks();
 }
 
-int CompatMidiRender::slideTicks(Chord* chord)
+int slideTicks(Chord* chord)
```
Comment (RomanPudashkin) [inline src/engraving/compat/midi/compatmidirender.cpp:1182]: These functions should be static

File: src/engraving/tests/midirenderer_tests.cpp [hunk 4]
```diff
@@ -279,6 +293,21 @@ TEST_F(MidiRenderer_Tests, graceOnBeatAndGlissando)
 
     EventsHolder events = renderMidiEvents(u"grace_on_beat_and_glissando.mscx");
 
+    // LOGE() << "@# events size = " << events.size();
```
Comment (RomanPudashkin) [inline src/engraving/tests/midirenderer_tests.cpp:296]: Do you still need this?

File: src/engraving/compat/midi/compatmidirenderinternal.cpp [hunk 5]
```diff
@@ -572,23 +746,32 @@ static void collectNote(EventsHolder& events, const Note* note, const CollectNot
...
-            break;
+    if (note->configuration()->useStretchedBends()) {
+        if (StretchedBend* stretchedBend = note->stretchedBend()) {
+            collectBend(stretchedBend->pitchValues(), stretchedBend->staffIdx(), noteChannel, tick1, tick1 + getPlayTicksForBend(
+                            note).ticks(), pitchWheelRenderer, noteEffect);
         }
+    } else {
+        GuitarBend* bendFor = note->bendFor();
+        if (bendFor) {
+            collectGuitarBend(note, noteChannel, tick1, noteParams.graceOffsetOn, noteParams.previousChordTicks, pitchWheelRenderer,
```
[the remark this answers is not in the document]
Comment (alexpavlov96) [reply src/engraving/compat/midi/compatmidirenderinternal.cpp:757]: any consequent combination of ties/bends are treated in 1 pass, and ignored if called again.
Rechecked, this combination also works. @abariska FYI

File: src/engraving/compat/midi/compatmidirender.cpp (hunk 2 above)
Comment (RomanPudashkin) [reply src/engraving/compat/midi/compatmidirender.cpp:507]: yeah, with true :) there are a bunch of similar flags in other places, and they are all inverted :)

File: src/engraving/compat/midi/compatmidirenderinternal.cpp [hunk 6]
```diff
@@ -304,6 +307,149 @@ static void collectVibrato(int channel,
...
+        int duration = 0;
+        if (bendFor && bendFor->type() == GuitarBendType::GRACE_NOTE_BEND) {
+            duration = (previousChordTicks == -1) ? GRACE_BEND_DURATION : std::min(previousChordTicks / 2, GRACE_BEND_DURATION);
+        } else {
+            duration = note->chord()->actualTicks().ticks();
+        }
+        curPitchBendSegmentEnd = curPitchBendSegmentStart + duration;
+
+        if (bendFor) {
+            Note* endNote = bendFor->endNote();
```
Comment (RomanPudashkin) [inline src/engraving/compat/midi/compatmidirenderinternal.cpp:386]: Please make all pointers const if the objects they point to should not be modifiable

**Review by RomanPudashkin — APPROVED** [review]

**Review by mikekirin — APPROVED** [review]

**Review by abariska — APPROVED** [review]
There're some minor issues but they will be reported separately

**Pull request merged**